group-coordinator/src/test/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntimeTest.java

```diff
@@ -0,0 +1,803 @@
+        when(loader.load(TP, coordinator)).thenReturn(future);
+
+        // Schedule the loading.
+        runtime.scheduleLoadOperation(TP, 10);
+
+        // Getting the context succeeds and the coordinator should be in loading.
+        CoordinatorRuntime<MockCoordinator, String>.CoordinatorContext ctx = runtime.contextOrThrow(TP);
+        assertEquals(CoordinatorRuntime.CoordinatorState.LOADING, ctx.state);
+        assertEquals(10, ctx.epoch);
+        assertEquals(coordinator, ctx.coordinator);
+
+        // When the loading completes, the coordinator transitions to active.
+        future.complete(null);
+        assertEquals(CoordinatorRuntime.CoordinatorState.ACTIVE, ctx.state);
+        assertEquals(10, ctx.epoch);
+
+        // Loading with a previous epoch is a no-op. The coordinator stays
+        // in active state with the correct epoch.
+        runtime.scheduleLoadOperation(TP, 0);
+        assertEquals(CoordinatorRuntime.CoordinatorState.ACTIVE, ctx.state);
+        assertEquals(10, ctx.epoch);
+    }
+
+    @Test
+    public void testScheduleLoadingAfterLoadingFailure() {
+        MockCoordinatorLoader loader = mock(MockCoordinatorLoader.class);
+        MockCoordinatorBuilderSupplier supplier = mock(MockCoordinatorBuilderSupplier.class);
+        MockCoordinatorBuilder builder = mock(MockCoordinatorBuilder.class);
+        MockCoordinator coordinator = mock(MockCoordinator.class);
+
+        CoordinatorRuntime<MockCoordinator, String> runtime =
+            new CoordinatorRuntime.Builder<MockCoordinator, String>()
+                .withLoader(loader)
+                .withEventProcessor(new MockEventProcessor())
+                .withPartitionWriter(new MockPartitionWriter())
+                .withCoordinatorBuilderSupplier(supplier)
+                .build();
+
+        when(builder.withSnapshotRegistry(any())).thenReturn(builder);
+        when(builder.build()).thenReturn(coordinator);
+        when(supplier.get()).thenReturn(builder);
+        CompletableFuture<Void> future = new CompletableFuture<>();
+        when(loader.load(TP, coordinator)).thenReturn(future);
+
+        // Schedule the loading.
+        runtime.scheduleLoadOperation(TP, 10);
+
+        // Getting the context succeeds and the coordinator should be in loading.
+        CoordinatorRuntime<MockCoordinator, String>.CoordinatorContext ctx = runtime.contextOrThrow(TP);
+        assertEquals(CoordinatorRuntime.CoordinatorState.LOADING, ctx.state);
+        assertEquals(10, ctx.epoch);
+        assertEquals(coordinator, ctx.coordinator);
+
+        // When the loading fails, the coordinator transitions to failed.
+        future.completeExceptionally(new Exception("failure"));
+        assertEquals(CoordinatorRuntime.CoordinatorState.FAILED, ctx.state);
+
+        // Verify that onUnloaded is called.
+        verify(coordinator, times(1)).onUnloaded();
+
+        // Create a new coordinator.
+        coordinator = mock(MockCoordinator.class);
```

Comment:
Do we need to create a new coordinator here? We don't seem to verify that it is different.

Reply:
I do this to ensure that a new coordinator is created in this case. We verify the new value a few line below: `assertEquals(coordinator, ctx.coordinator);`.

Reply:
Ah I see. When I saw `assertEquals(coordinator, ctx.coordinator)` before I was confused because that would be true with the old coordinator. But we are testing that the ctx is updating.